sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -267,10 +269,22 @@ class FileStreamSource(
     val logOffset = FileStreamSourceOffset(end).logOffset
 
     sourceCleaner.foreach { cleaner =>
-      val files = metadataLog.get(Some(logOffset), Some(logOffset)).flatMap(_._2)
-      val validFileEntities = files.filter(_.batchId == logOffset)
-      logDebug(s"completed file entries: ${validFileEntities.mkString(",")}")
-      validFileEntities.foreach(cleaner.clean)
+      sourceHasMetadata match {
+        case Some(true) if !warnedIgnoringCleanSourceOption =>
+          logWarning("Ignoring 'cleanSource' option since source path refers to the output" +
+            " directory of FileStreamSink.")
+          warnedIgnoringCleanSourceOption = true
+
+        case Some(false) =>
+          val files = metadataLog.get(Some(logOffset), Some(logOffset)).flatMap(_._2)
+          val validFileEntities = files.filter(_.batchId == logOffset)
+          logDebug(s"completed file entries: ${validFileEntities.mkString(",")}")
+          validFileEntities.foreach(cleaner.clean)
+
+        case _ =>
+          logWarning("Ignoring 'cleanSource' option since Spark hasn't figured out whether " +
```

Comment:
I just put logWarning here - I was about to throw IllegalStateException here since it doesn't sound feasible to have some files from commit() and FileStreamSource still cannot decide, but there might be some edge-case so avoided being aggressive here.

Reply:
How about throwing an UnsupportedOperationException here: https://github.com/apache/spark/blob/ba2bc4b0e0eea0c1b6732a18cb20e61e4f693156/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala#L205

Reply:
The only "odd" case I can imagine to reach here is, 

1) the query executed which wrote the commit log of the last batch and stopped before writing the offset for next batch.
2) the query is restarted, and `constructNextBatch` is called.
3) somehow the source files are all deleted between 1) and 2), hence FileStreamSource doesn't see any file and cannot decide when `fetchAllFiles` is called.
4) `constructNextBatch` will call `commit` for previous batch the query executed before.

It's obviously very odd case as the content of source directory are modified (maybe) manually which we don't support the case (so throwing exception would be OK), but I'm not fully sure there's no another edge-cases.

Btw, where do you recommend to add the exception? L287, or L205?

Reply:
Also not yet see which place is the suggestion refers to.

> L205, I'm not sure I follow

+1

L287: As I see this is more or less the `should never happen` case. The question is whether we can consider edge cases which may hit this. If we miss a valid case and we're throwing exception here we may block a query to start.

Reply:
> 3\. somehow the source files are all deleted between 1) and 2)

This should be a user error.

My general point is we should make sure the data files and the metadata in `_spark_metadata` are consistent and we should prevent from cleaning up data files that are still tracked. Logging a warning without really deleting files is a solution, however, most of users won't be able to notice this warning from their logs. Hence we should detect this earlier. There is already a variable `sourceHasMetadata` tracking whether the source is reading from a file stream sink or not. We can check the options and throw an exception when flipping it. What do you think?

Reply:
Ah OK I guess I got your point now. I'm also in favor of being "fail-fast" and the suggestion fits it. Thanks! Just updated.